### PR TITLE
Fixes #30957 - External IPAM improved error handling and doc updates

### PR DESCRIPTION
**Pull request by grizzthedj**
The External IPAM plugin(https://github.com/grizzthedj/smart_proxy_ipam) implementation has been refactored to separate the API from provider implementations, to increase stability, and make it easier to add new IPAM providers. 

As such, the data contract between Foreman and this plugin has been simplified, and this patch addresses this. 

Netbox has also been added as a new IPAM provider, and is fully functional. This will be released in coming days as smart_proxy_ipam version 0.1.0.

### CODE REVIEW

**Comment by ekohl** [comment]
> Netbox has also been added as a new IPAM provider, and is fully functional. This will be released in coming days as smart_proxy_ipam version 0.1.0.

I have been thinking about including it in smart-proxy core. At least the generic plugin, even if there is no provider. Providers can then be implemented in specific gems. This would be in line with the policy where all functionality that's in Foreman core is also in Smart Proxy core.

With the refactored base, I think we're much closer to being able to do that.

**Comment by ezr-ondrej** [comment]
@lzap do you wanna take this?

File: app/models/subnet.rb
Comment (lzap) [inline app/models/subnet.rb:301]: Just checking, no change except line order?

File: lib/proxy_api/external_ipam.rb
Comment (lzap) [inline lib/proxy_api/external_ipam.rb]: This is actually good practice, having root element enables you to extend the API without breaking clients! I highly suggest the opposite - to review all calls to actually have a root element called "data" ;-)

I mentioned this in your smart proxy PR the other day.

File: lib/proxy_api/external_ipam.rb
Comment (lzap) [inline lib/proxy_api/external_ipam.rb]: Returning empty elements like arrays or hashes is also a good practice I would not be getting rid of. Exactly vice versa!

File: lib/proxy_api/external_ipam.rb
Comment (lzap) [inline lib/proxy_api/external_ipam.rb]: This method should actually expect the "data" element and clean it out for caller. The code was better before.

**Review by lzap — CHANGES_REQUESTED** [review]
I missed this PR, sorry! Thanks for headsup @ezr-ondrej 

I think this is a change for the worse, root elements are good. Can you actually do the opposite? Review all API calls and make sure they include "data" element, strip it out in this class so callers don't need to know.

**Review by lzap — DISMISSED** [review]
This is okay, but the title is a bit misleading. This is nowhere near to "refactoring", this is a documentation update with some error handling improvements. Can you reword the title a bit?

**Review by lzap — APPROVED** [review]
All of this are documentation updates or slight changes in error handling. Can we merge this for 2.3 @tbrisker please?

**Comment by tbrisker** [comment]
ok to test
Let's first merge it to nightly post-branching then consider pulling it into 2.3-stable

**Comment by tbrisker** [comment]
[test foreman]

**Comment by lzap** [comment]
I haven't realized we have branched already. To explain, @grizzthedj preferred this in 2.3 because of a big refactoring of the plugin which could be released in sync with 2.3. If we can do it, that would be great. Thanks!

**Comment by tbrisker** [comment]
@lzap we are currently in the branching process so I'd rather not merge until we're done with it, afterward we can merge and cherry-pick to 2.3-stable. ping me in a day or two and we'll take care of it.

**Comment by lzap** [comment]
Thank you, 2.3 was branched off, merged for 2.4. If you need this in 2.3 @grizzthedj let me know and I can file a PR.

**Comment by grizzthedj** [comment]
@lzap Ideally I would like to get it in 2.3, this way people can start using the Netbox integration sooner.

**Comment by tbrisker** [comment]
2.3 - cee4bf4b2ebbc9f29eefb834940ed18166e3668f

**Pull request merged**
